### PR TITLE
Port LatticeField to v3

### PR DESCRIPTION
**Pull request by tcmoore3**
## Description

The primary purpose of this PR is to port the lattice field to the v3 API. The current implementation only allows a single external field to be applied to the systems. This limitation will be addressed in a future PR.

## Motivation and context

This PR covers a portion of #878. 

## How has this been tested?

I have tested locally that particles remain in their reference positions and orientations with strong springs, and that their fluctuations grow in size with decreasing spring constants. I will add unit tests.

## Change log

```
Changed hoomd.hpmc.field.lattice_field to hoomd.hpmc.field.LatticeField
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/hpmc/ExternalFieldLattice.h
Comment (tcmoore3) [inline hoomd/hpmc/ExternalFieldLattice.h:54]: I couldn't get this to compile

File: hoomd/hpmc/ExternalFieldLattice.h
Comment (tcmoore3) [inline hoomd/hpmc/ExternalFieldLattice.h:276]: I left the box stuff commented out here. We either need to remove this stuff or allow this field with variable box sizes (which will take a little more thought about the right way to do).

File: hoomd/hpmc/ExternalFieldLattice.h
Comment (tcmoore3) [inline hoomd/hpmc/ExternalFieldLattice.h:420]: This commented out code will leave once I figure out the the particle change number signal bit.

File: hoomd/hpmc/field.py
Comment (tcmoore3) [inline hoomd/hpmc/field.py:40]: This class will need to inherit from a different (probably new) class that all of the external fields inherit from.

File: hoomd/hpmc/field.py
Comment (tcmoore3) [inline hoomd/hpmc/field.py:89]: Do we need this?

File: hoomd/hpmc/ExternalFieldLattice.h
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldLattice.h]: This works now.

File: hoomd/hpmc/ExternalFieldLattice.h
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldLattice.h]: I figured that other bit out. This is gone now.

**Comment by tcmoore3** [comment]
Some to-dos:

- [ ] Update (write?) unit tests for `LatticeField` class
- [ ] Decide what to do when the box can change and a `LatticeField` is active

File: hoomd/hpmc/field.py
Comment (tcmoore3) [reply hoomd/hpmc/field.py]: Done.

File: hoomd/hpmc/field.py
Comment (tcmoore3) [reply hoomd/hpmc/field.py]: No.

File: hoomd/hpmc/field.py
Comment (joaander) [inline hoomd/hpmc/field.py:72]: Add a more complete description of how this is computed with respect to the given symmetries.

File: sphinx-doc/module-hpmc-field.rst
Comment (joaander) [inline sphinx-doc/module-hpmc-field.rst:11]: I find `hpmc.field.LatticeField` both redundant and not informative. How about `hpmc.field.Harmonic` or `hpmc.field.Einstein`? I'm open to other suggestions.

File: hoomd/hpmc/ExternalFieldLattice.h
Comment (joaander) [reply hoomd/hpmc/ExternalFieldLattice.h:278]: I say remove. Future developers can add scaling back in the correct way if they find a strong use case for it.

File: hoomd/hpmc/ExternalFieldLattice.h
Comment (joaander) [inline hoomd/hpmc/ExternalFieldLattice.h:367]: Remove `scale` here too and in the other methods.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! This is looking good. I've requested some changes.

Also, be careful with `git commit -a`, it appears you committed a different version of the tutorials. Revert with:
```
cd sphinx-doc/tutorial
git checkout 92038ecc811ca450acd3c2912eb818c4ebece06b
cd ..
git add tutorial
git commit
```

File: hoomd/hpmc/field.py
Comment (tcmoore3) [reply hoomd/hpmc/field.py:52]: Just to be clear, this docstring needs the `r"""` because of the backslashes in the equations, right?

File: hoomd/hpmc/field.py
Comment (tcmoore3) [reply hoomd/hpmc/field.py:71]: Yes, definitely.

File: hoomd/hpmc/field.py
Comment (joaander) [reply hoomd/hpmc/field.py:52]: Perhaps, but not necessarily. In may places in the docs, I need to use `"""` and escape every backslash so that the single backslash character can be used for line continuation to correctly format sphinx docs while meeting the flake8 checks.

File: hoomd/hpmc/field.py
Comment (tcmoore3) [reply hoomd/hpmc/field.py]: I made that change.

File: hoomd/hpmc/field.py
Comment (tcmoore3) [reply hoomd/hpmc/field.py]: Done.

File: sphinx-doc/module-hpmc-field.rst
Comment (tcmoore3) [reply sphinx-doc/module-hpmc-field.rst]: I changed this to `hpmc.field.Harmonic`, and also updated the C++ class to match.

File: hoomd/hpmc/field.py
Comment (tcmoore3) [reply hoomd/hpmc/field.py]: I changed it to include the factor of 1/2.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (joaander) [inline hoomd/hpmc/ExternalFieldHarmonic.h:44]: Reading the code, it appears that the user is required to include the identity in the list. This should be documented.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (joaander) [inline hoomd/hpmc/ExternalFieldHarmonic.h:181]: If this returns a copy, why does it make it not writeable?
@b-butler does this need any special handling to interoperate with #1178?

File: hoomd/hpmc/field.py
Comment (joaander) [inline hoomd/hpmc/field.py:87]: Encode this description in the math equation succinctly. Something like:
```\frac{1}{2} k_{rot} \cdot \min_j \left[ (q_i-q_{0,i} \cdot q_{\mathrm{symmetry},j})^2 \right]```
Then the text description becomes: "where q_{\mathrm{symmetry}} is the given set of symmetric orientations."

File: hoomd/hpmc/field.py
Comment (joaander) [inline hoomd/hpmc/field.py:88]: Need an `Attributes:` section that documents the `param_dict` entries. It can mostly be copied and pasted from `Args:`.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! Some additional changes are needed.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldHarmonic.h:44]: Added.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldHarmonic.h:181]: Deferring to @b-butler

File: hoomd/hpmc/field.py
Comment (tcmoore3) [reply hoomd/hpmc/field.py:87]: Yes, beautiful. I couldn't figure out how to write this succinctly because I was unaware of the `\min` LaTeX command. Changed.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (b-butler) [reply hoomd/hpmc/ExternalFieldHarmonic.h:181]: The reason this is not writable (I believe this is a copy paste from another updater) is because it would be deceiving as to what is happening. It is the copy-modify-write problem again. Given the fact that NumPy is not part of the Python standard library I did not create an array class that continually updates. It is possible though.

Comment (joaander) [reply hoomd/hpmc/ExternalFieldHarmonic.h:181]: That makes sense. This is attempting to give some notion of `const` or immutability with the read/modify/write mechanism.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (joaander) [inline hoomd/hpmc/ExternalFieldHarmonic.h:398]: Add a note in the documentation that the rotational energy is zero for non-orientable particles (such as `hpmc.integrate.Sphere`).

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (joaander) [inline hoomd/hpmc/ExternalFieldHarmonic.h:313]: Would users be interested in accessing energy components separately in `energy_translational` and `energy_rotational`? If so, we could add these properties. Keep the `energy` property that computes the total.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks for the fast turnaround. There are just a few more updates needed. 
In addition to the inline comments, `Harmonic` should live somewhere under `hpmc.external`. Let's discuss on Slack.

The suggestion to add separate properties for translational and rotational energy is optional, but would be nice.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldHarmonic.h:313]: Yes, I think users would be interested in something like that. I haven't worked out the details but I imagine this would be useful for getting rotational and translational contributions to the free energy somehow. And it's easy enough to add, so I'll put it in.

File: hoomd/hpmc/external/field.py
Comment (joaander) [inline hoomd/hpmc/external/field.py:51]:
```suggestion
    r"""Restrain particle positions and orientations with harmonic springs.
```
As written, the `noqa` is part of the docstring. We need to format the docstring appropriately so that it obeys the linter rules.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (joaander) [inline hoomd/hpmc/ExternalFieldHarmonic.h:309]: Except that this isn't exported as a property with pybind. Typically the first sentence in a documentation comment describes what the method does: "Compute the total external energy on the system."

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (joaander) [inline hoomd/hpmc/ExternalFieldHarmonic.h]: I prefer not to introduce boolean parameters that get must get passed down the chain. Instead implement `getEnergyTranslational` and `getEnergyRotational` as separate methods.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (joaander) [reply hoomd/hpmc/ExternalFieldHarmonic.h:399]: On second review, I noticed some issues here. The method only the `calcE_rot(const unsigned int& index, const Shape& shape)` knows whether the shape is orientable. I haven't search the whole codebase, but it appears this method is not even called. `getEnergy`, `energydiff`, and `calculateDeltaE` all call the other signature.

1) Remove `calcE_rot` and see if it compiles, that would verify it is not used.
2) Update the docs to remove the mention of a zero rotational energy. Users are responsible for setting `k_rotational=0` for non-orientable particles.

File: hoomd/hpmc/external/field.py
Comment (joaander) [inline hoomd/hpmc/external/field.py:201]:
```suggestion
            :math:`[\\mathrm{energy}]`. 
```
Don't include noqa in docstrings.

File: hoomd/hpmc/external/field.py
Comment (joaander) [inline hoomd/hpmc/external/field.py:220]:
```suggestion
    r"""Manage multiple external fields.
```
Don't include noqa in docstrings.

I notice there are many more docstrings that include noqa, please fix all of them.

File: hoomd/hpmc/external/field.py
Comment (joaander) [inline hoomd/hpmc/external/field.py:312]: `wall` doesn't belong in the `field.py` file. You can create a `wall.py` to keep the old v2 code for now and add it to the exceptions for pre-commit and flake8 instead of specifying noqa on every line.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! There are more updates needed before merging.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (b-butler) [inline hoomd/hpmc/ExternalFieldHarmonic.h:81]: You error if `N_particles != m_pdata->getNGlobal()` so no need for the min call.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (b-butler) [inline hoomd/hpmc/ExternalFieldHarmonic.h:88]:
```suggestion
            for (size_t i = 0; i < N_particles; i+=3)
                {
                this->m_reference_positions[i] = vec3<Scalar>(rawdata[i]],
                                                              rawdata[i + 1],
                                                              rawdata[i + 2]);
                }
            }
```

Comment (b-butler) [reply hoomd/hpmc/ExternalFieldHarmonic.h:88]: This can be applied to the other loops for this purpose.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (b-butler) [inline hoomd/hpmc/ExternalFieldHarmonic.h:265]:
```suggestion
    double calculateDeltaE(const Scalar4* const position_old_arg,
```

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (b-butler) [inline hoomd/hpmc/ExternalFieldHarmonic.h:366]: This is only used in box wap. Go ahead and create it there.

File: hoomd/hpmc/external/field.py
Comment (b-butler) [inline hoomd/hpmc/external/field.py:67]: Add note here that this must include the identity quaternion rather than a note below.

File: hoomd/hpmc/external/field.py
Comment (b-butler) [inline hoomd/hpmc/external/field.py:117]: You require double's in the signatures in C++ but use single precision here. That involves a needless copy. Change one or the other.

File: hoomd/hpmc/external/field.py
Comment (b-butler) [reply hoomd/hpmc/external/field.py:220]: Most of these are left overs from v2. For those I would recommend removal. Most of these should cause errors. For those that may be useful for v3, we can do as you suggest for walls and for the rest using noqa seems appropriate to not create work that will likely have to be redone.

For the classes and functions ported though, they should not have noqa.

File: hoomd/hpmc/pytest/test_external.py
Comment (b-butler) [inline hoomd/hpmc/pytest/test_external.py:17]: If we are only going to test one combination of parameters no need to make the test parametrized. If we wanted to test multiple combinations then that would make sense.

File: hoomd/hpmc/pytest/test_external.py
Comment (b-butler) [inline hoomd/hpmc/pytest/test_external.py:124]: This code appears in most tests. It would be a great canidate for a fixture that returns a simulation populated with the integrator and external potential.

File: hoomd/hpmc/pytest/test_external.py
Comment (b-butler) [inline hoomd/hpmc/pytest/test_external.py:144]: This should be unnecessary.

File: hoomd/hpmc/pytest/test_external.py
Comment (b-butler) [inline hoomd/hpmc/pytest/test_external.py:149]: This should be unnecessary.

**Review by b-butler — CHANGES_REQUESTED** [review]
The PR overall looks good. I have some comments to address mostly for code cleanness or clarity.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldHarmonic.h]: We resolved this in real-time.

File: hoomd/hpmc/external/field.py
Comment (tcmoore3) [reply hoomd/hpmc/external/field.py]: A lot of the `noqa`s are indeed in docs for classes that have not been ported over to v3 yet. I removed several of those classes and there are no more `noqa`s in the classes relevant to this PR.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldHarmonic.h:81]: Good catch! Changed.

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldHarmonic.h:370]: Do you just mean move it to the line before the `box.wrap()` line?

File: hoomd/hpmc/external/field.py
Comment (tcmoore3) [reply hoomd/hpmc/external/field.py:118]: Changed.

File: hoomd/hpmc/pytest/test_external.py
Comment (tcmoore3) [reply hoomd/hpmc/pytest/test_external.py:16]: Added more sets of parameters to test.

File: hoomd/hpmc/pytest/test_external.py
Comment (tcmoore3) [reply hoomd/hpmc/pytest/test_external.py:123]: Feel free to clean this up if desired.

File: hoomd/hpmc/pytest/test_external.py
Comment (tcmoore3) [reply hoomd/hpmc/pytest/test_external.py:148]: Even if the energy property has `requires_run=True`?

File: hoomd/hpmc/pytest/test_external.py
Comment (tcmoore3) [reply hoomd/hpmc/pytest/test_external.py:143]: Even if the energy property has `requires_run=True`?

File: hoomd/hpmc/ExternalFieldHarmonic.h
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldHarmonic.h]: > This can be applied to the other loops for this purpose.

...with appropriate modifications of course (:

**Comment by tcmoore3** [comment]
@joaander @b-butler This is ready for another review. My preference is to go ahead and merge this in if there are real issues and only cosmetic issues. Feel free to go ahead and make any style changes you prefer.

**Review by joaander — APPROVED** [review]
Thanks!

I pushed my suggested documentation style changes directly to the branch.

File: hoomd/hpmc/pytest/test_external.py
Comment (b-butler) [reply hoomd/hpmc/pytest/test_external.py:158]: Yes because run is called above.

**Comment by joaander** [comment]
Tests are now failing: https://github.com/glotzerlab/hoomd-blue/runs/4870480987?check_suite_focus=true#step:6:6008

I fixed the pre-commit failure on master, you can merge the fix in.

**Review by b-butler — APPROVED** [review]

**Pull request merged**
